### PR TITLE
[Fix #5223] Skip date format strings in Style/FormatStringToken

### LINKED ISSUE #5223: Style/FormatStringToken unannotated - false negative for Date/Time format
Rubocop 0.52.0 [introduced](https://github.com/bbatsov/rubocop/commit/b32b260df17e6370bff872ef4a2b551880f3411e) check for unannotated style, and now it throws error on every timestamp expression with default configuration. It makes sense for sting interpolation, but it does not for string to date conversion pattern.

## Expected behavior
... not to warn about `'%Y-%m-%dT%H:%M:%S.%L%z'`

## Actual behavior

It throws `Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).`

## Steps to reproduce the problem
Code that causes warning:
```ruby
Time.strptime('2017-10-13T12:56:33.000Z', '%Y-%m-%dT%H:%M:%S.%L%z')
```

## RuboCop version
```
$ rubocop -V
0.52.0 (using Parser 2.4.0.2, running on ruby 2.4.2 x86_64-darwin17)
```

### PR DESCRIPTION
**Pull request by nattfodd**
The https://github.com/bbatsov/rubocop/pull/5068 PR introduced new style for `Style/FormatStringToken` - `unannotated` with it's own regex. This regex matches date format strings and urls, and therefore marks them as offence.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## master (unreleased)
 
+* [#5223](https://github.com/bbatsov/rubocop/issues/5223): Do not check date format strings with Style/FormatStringToken. ([@nattfodd][])
```
Comment (pocke) [inline CHANGELOG.md]: Can you add a header for the change?

For example

```markdown
## master (unreleased)

### Bug fixes

* [#5223](...) ...

```

File: lib/rubocop/cop/style/format_string_token.rb [hunk 2]
```diff
@@ -137,6 +138,13 @@ def slice_source(source_range, new_begin, new_end)
             new_end
           )
         end
+
+        def date_format?(node)
+          parent_node = node.parent
+          return unless parent_node
```
Comment (Drenmi) [inline lib/rubocop/cop/style/format_string_token.rb]: This is missing a check for the parent node type. (Should be `send`.) 🙂

**Comment by bbatsov** [comment]
I'm guessing the assumption with this fix is that it's unlikely for someone to assign a date-formatting string to a variable or a constant, right?

**Comment by Nattfodd** [comment]
@bbatsov I'm not sure if it's there's another way to check such things...

**Comment by mvz** [comment]
Since there are a lot of problems with false positives for this cop, why not check at the point where `format` is invoked? E.g.:

```ruby
format(some_string, foo, bar) # bad
format(other_string, foo: foo, bar: bar) # good
```

This fixes problems with date formats, URLs, etc. in one go.

**Comment by Drenmi** [comment]
> Since there are a lot of problems with false positives for this cop, why not check at the point where format is invoked?

The reason is that there are more than two styles, some of which can't be distinguished at the point of formatting. 😕

**Comment by mvz** [comment]
> The reason is that there are more than two styles, some of which can't be distinguished at the point of formatting.

True, but at least the FormatStringToken cop can stop worrying about strings that look like `%s%d%m%i` and only check `%{foo}` vs `%<foo>s`.

**Comment by kddeisz** [comment]
Womp...

```ruby
lib/humidifier/aws_adapters/sdkv2.rb:6:28: C: Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
      TIME_FORMAT = '%Y-%m-%d-%H-%M-%S'.freeze
```

Don't assume they're not being assigned to constants.

**Comment by benhutton** [comment]
i wonder if there are checks like "tokens are > N% of the string" that would give really high confidence that we're looking at a time format?

**Comment by mvz** [comment]
@benhutton I have strings that contain nearly only tokens but are not time formats.

**Comment by Nattfodd** [comment]
I've updated pull requests reversing approach. The idea is: if somewhere `:%` or `:sprintf` methods are called, we check those strings with `:unannotated` style. Otherwise - ignore them.

**Comment by mvz** [comment]
@Nattfodd don't forget about `format`.

**Comment by Nattfodd** [comment]
@mvz thanks, added

**Comment by bbatsov** [comment]
> I've updated pull requests reversing approach. The idea is: if somewhere :% or :sprintf methods are called, we check those strings with :unannotated style. Otherwise - ignore them.

My only concern so far is that it seems we have an inconsistent approach to checking the format strings for different styles. I think regardless of the style in question we should check just the params to those methods.

**Comment by nattfodd** [comment]
> My only concern so far is that it seems we have an inconsistent approach to checking the format strings for different styles. I think regardless of the style in question we should check just the params to those methods.

I left `annotated` & `template` styles untouched, because their regexes are pretty specific, so complaints about false positives are barely happen, and those styles are pretty nice and straightforward. On the other hand, `unannotated` style has a very wide regex, which leads to a lot of false positives.

The fix in this PR isn't the best one possible, there is still a room for false negatives in some cases (for example, string format is declared in one file, and it is used in another one - we gonna miss that and do not warn user). So by making an exclusion for `unannotated` style, I leave the good `annotated` & `template` styles in their original state, and only restrict `unannotated` one to be more selective.

**Comment by bbatsov** [comment]
This makes sense, but those design choices should be explained somewhere - e.g. in the cop's class-level documentation. 

Once you do this you can also rebase on top of the current `master`.

**Comment by nattfodd** [comment]
Rebase: done
Class commentary: done

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**